src/iterset.ts

```diff
@@ -50,29 +79,94 @@ export function iterSet <M extends QueryMethodTypes> (
-    if (!iterResult.accept(new Date(zonedDate.getTime()))) break
-  }
+  let recurringObjectSearchTree = new BinarySearchTree<number, {
+    iterator: IterableIterator<Date>,
+    key: Date,
+    dataValue: Date | RRule
+  }>()
 
-  _rrule.forEach(function (rrule) {
-    iter(iterResult, rrule.options)
+  const filteredRDate = _rdate.filter((date) => {
+    const zonedDate = new DateWithZone(date, tzid).rezonedDate()
+    return iterResult.accept(new Date(zonedDate.getTime()))
   })
 
-  const res = iterResult._result
-  dateutil.sort(res)
-  switch (iterResult.method) {
-    case 'all':
-    case 'between':
-      return res as IterResultType<M>
-    case 'before':
-      return ((res.length && res[res.length - 1]) || null) as IterResultType<M>
-    case 'after':
-    default:
-      return ((res.length && res[0]) || null) as IterResultType<M>
+  if (filteredRDate.length) {
+    // Assumes that the array is already sorted
+    const dateIterator = arrayToIterator<Date>(filteredRDate)
+    let firstValueAccepted = dateIterator.next().value
+
+    if (firstValueAccepted) {
+      recurringObjectSearchTree.insert(firstValueAccepted.valueOf(), {
+        iterator: dateIterator,
+        key: firstValueAccepted,
+        dataValue: firstValueAccepted
+      })
+    }
+
+  }
+
+  for (const rrule of _rrule) {
+    const rruleIterator = iter(iterResult, rrule.options)
+    const firstValue = rruleIterator.next().value
+
+    if (!firstValue) continue
+    recurringObjectSearchTree.insert(firstValue.valueOf(), {
+      iterator: rruleIterator,
+      key: firstValue,
+      dataValue: rrule
+    })
+  }
+
+  const isBefore = iterResult.method === 'before'
+  const isAllOrBetween = iterResult.method === 'all' || iterResult.method === 'between'
+  let lastResult: IteratorSetReturn | undefined
+
+  while (recurringObjectSearchTree.count() > 0) {
+    // Find the smallest date
+    const min = recurringObjectSearchTree.min()
+
+    // Remove it from the tree
+    recurringObjectSearchTree.remove(min.getKey())
+    const {
+      iterator,
+      key,
+      dataValue
+    } = min.getValue()
+
+    // Yield the date with a reference to the RRule that generated it
+    const result: IteratorSetReturn = [key, dataValue]
+
+    if (isAllOrBetween) {
+      yield result
+    } else if (isBefore) {
+      // Return the last element
+      lastResult = result
+    } else {
+      yield result
+      return
+    }
+
+    // If we are after, we only want to return the first date
+
+    // Advance to the next date if it exists
+    const nextValue = iterator.next().value
+    if (nextValue === undefined) continue
+
+    // Reinsert the iterator with the next value into the tree
+    recurringObjectSearchTree.insert(nextValue.valueOf(), {
+      iterator: iterator,
+      key: nextValue,
+      dataValue: dataValue
+    })
+
+  }
+
+  if (isBefore === true && lastResult !== undefined) {
+    yield lastResult
```

Comment:
This chunk here is main piece of added logic that:

Inserts rrules/rdates into a binary search tree sorted by the time stamp of the first event. Every time we pull out the earliest event, emit the result, generate a its next event, and insert it into the BTREE in sorted order. 

Repeat, and all events will be emitted in sorted order.